## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [69]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [70]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [71]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [72]:
df[['unit_price','unit_cost']]

unit_price  unit_cost
0         1096.80     762.71
1           24.95      15.09
2           13.52       7.86
3           36.42      24.85
4          195.10     108.17
...           ...        ...
54634       83.13      51.60
54635      206.59     175.46
54636       65.66      57.31
54637        1.48       1.22
54638        3.01       1.61

[54639 rows x 2 columns]

In [73]:
# Use this space to do any additional research
# and familiarize yourself with the data.

# find all clients who's name starts wit g
df[df['first'].str.startswith('G')]


first      last                              job  \
9        Greg    Garcia               Secretary, company   
24      Glenn    Garcia                 Textile designer   
102    George  Castillo                   Engineer, land   
250      Gary    Pierce  Chartered management accountant   
273    Gordon   Webster                           Gaffer   
...       ...       ...                              ...   
54180    Gary    Pierce  Chartered management accountant   
54290   Glenn    Garcia                 Textile designer   
54328  Gordon   Webster                           Gaffer   
54608  Gloria   Navarro           Audiological scientist   
54623  Gloria   Navarro           Audiological scientist   

                        phone                                      email  \
9          (231)342-7216x1421                        ggarcia@bentley.biz   
24               573.981.6374          determinegarcia5685@velasquez.com   
102              865.902.1326  developmentcastillo1825@smith-brennan.com   
250    001-718-376-5473x39181               institutiongary@thompson.com   
273    001-324-901-3685x04857            glisten724@williams-morgan.info   
...                       ...                                        ...   
54180  001-718-376-5473x39181               institutiongary@thompson.com   
54290            573.981.6374          determinegarcia5685@velasquez.com   
54328  001-324-901-3685x04857            glisten724@williams-morgan.info   
54608           (481)651-9381                 navarrogloria2626@lane.com   
54623           (481)651-9381                 navarrogloria2626@lane.com   

       client_id  order_id  order_date  order_week  order_year  \
9          58789   9665852  2023-04-14          15        2023   
24         33804   6024379  2023-01-26           4        2023   
102        19206   9461732  2023-05-28          21        2023   
250        33921   8420525  2023-01-14           2        2023   
273        29026   4677030  2023-04-19          16        2023   
...          ...       ...         ...         ...         ...   
54180      33921   9359347  2023-04-19          16        2023   
54290      33804   8921120  2023-03-13          11        2023   
54328      29026   4677030  2023-04-19          16        2023   
54608      89110   2707963  2023-05-12          19        2023   
54623      89110   9012482  2023-02-05           5        2023   

              item_id     category     subcategory  unit_price  unit_cost  \
9      DUU85042-87-5E    furniture         seating       66.88      54.15   
24     XZZ76042-95-6X  consumables  paper products        5.17       3.53   
102    MEX67117-45-3X        decor        wall art       23.40      21.53   
250    XED19763-96-7Z    furniture          stands       62.72      43.27   
273    CUZ68886-63-5E        decor            misc      126.16     106.09   
...               ...          ...             ...         ...        ...   
54180  BAA43653-15-9C        decor          plants      184.24     116.47   
54290  DEX15141-48-9B    furniture         seating      174.61      96.01   
54328  MUZ81503-74-3D    furniture          tables       65.90      58.86   
54608  CUU90642-03-7C    furniture         seating       33.41      26.82   
54623  EUE38901-78-7D    furniture           desks       90.81      76.71   

       unit_weight   qty  line_number  
9             2.47    78            1  
24            6.90  2163            2  
102           3.03    54            0  
250           0.83    28            0  
273           5.94   428            2  
...            ...   ...          ...  
54180         4.25    34            5  
54290         1.02   512            0  
54328         2.61   203            6  
54608        29.12   566            6  
54623         5.36   279            2  

[702 rows x 18 columns]

In [74]:
# What three item categories had the most entries?
# top_3 = df['category'].value_counts().head(3)
# top_3
df['category'].value_counts().head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [75]:
# Gathering the first and last names of all clients who's first name is Christopher

top_sub = df[df['first'] == "Christopher"]['last'].value_counts()
top_sub

last
Rodriguez    172
Hobbs        168
Hanson       115
Ortiz         82
Roberts       73
Hill          53
Rice          52
Bailey        49
Wilkinson     48
Johnson       44
Carter        41
Rhodes        35
Hawkins       31
Floyd         24
Wright        23
Vega          21
Arias         18
Gay           17
Herrera       14
Hernandez     10
Name: count, dtype: int64

In [76]:
# For the category with the most entries,
# which subcategory had the most entries?
top_category = df['category'].value_counts().idxmax()
top_sub = df[df['category'] == top_category]['subcategory'].value_counts().head(1)
top_sub


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [77]:
# Which five clients had the most entries in the data?
top_five_clients = df['client_id'].value_counts().head(5)
top_five_clients



client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [78]:
# Store the client ids of those top 5 clients in a list.
top_five_clients_list = top_five_clients.index.tolist()


top_five_clients_list


[33615, 66037, 46820, 38378, 24741]

In [79]:
# How many total units (the qty column) did the
# client with the most entries order order?
top_client = top_five_clients_list[0]
top_client_units = df[df['client_id'] == top_client]['qty'].sum()
top_client_units

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [80]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df['subtotal'] = df['unit_price'] * df['qty']


In [81]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
def claculate_shipping(row):
    unit_weight = row['unit_weight']
    qty = row['qty']
    if unit_weight* qty > 50:
        return (unit_weight * qty) * 7
    else:
        return (unit_weight * qty) * 10
    
df['shipping'] = df.apply(claculate_shipping, axis=1)
df['shipping'].head()


0    5512.50
1     312.90
2     458.64
3     356.70
4    6500.20
Name: shipping, dtype: float64

In [82]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
df['total'] = df['subtotal'] + df['shipping']
df['total'] += (df['total'] * .0925)
df['total'].head()


0    131839.076250
1       914.258625
2      1077.117600
3      1543.571400
4     11364.403500
Name: total, dtype: float64

In [83]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['cost'] = df['unit_cost'] * df['qty'] + df['shipping']
df['cost'].head()


0    85597.05
1      629.79
2      765.18
3     1077.35
4     8663.60
Name: cost, dtype: float64

In [84]:
# Create a column for the profit of
# each line using line cost and line price
df['profit'] = df['total'] - df['cost']
df['profit'].head().round(2)


0    46242.03
1      284.47
2      311.94
3      466.22
4     2700.80
Name: profit, dtype: float64

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [85]:
df[df['order_id']==2742071]['total'].sum().round(2)


152811.9

In [86]:
order_ids = [2742071, 2173913, 6128929]
order_totals = []

for order_id in order_ids:
    total_order = df[df['order_id'] == order_id]['total'].sum().round(2)
    order_totals.append({'Order ID': order_id, 'Total Price': total_order})

order_totals = pd.DataFrame(order_totals)
order_totals


Order ID  Total Price
0   2742071    152811.90
1   2173913    162388.72
2   6128929    923441.24

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [87]:
top_five_clients_names = df[df['client_id'].isin(top_five_clients_list)][['first','last']].drop_duplicates()
top_five_clients_names

first     last
5      Jessica    Reyes
137      Bryan    Myers
216     Angela  Everett
416     Kendra  Garrett
470  Alexandra    Young

In [92]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
top_five_clients = df[df['client_id'].isin(top_five_clients_list)]
top_five_clients

top_five_clients.groupby('client_id')['total'].sum().round(2)




client_id
24741    82268891.98
33615     8377308.57
38378    12906550.88
46820     9743794.32
66037    10259514.80
Name: total, dtype: float64

In [89]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.



In [90]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.



In [91]:
# Sort the updated data by "Total Profit" form highest to lowest
